File: QytetetJava/src/interfaztextualqytetet/ControladorQytetet.java
package interfaztextualqytetet;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Scanner;
import modeloqytetet.*;

/**
 *
 * @author aurelia
 */
public class ControladorQytetet {
    private Qytetet juego;
    private Jugador jugador;
    private Casilla casilla;
    private VistaTextualQytetet vista;
    private static final Scanner in = new Scanner (System.in);
    
    public void desarrolloJuego() throws IOException{
        boolean bancarrota = false;
        boolean tienePropietario;
        
        while (!bancarrota){
            vista.mostrar("Es el turno de " + jugador.getNombre() + "\n");
            casilla = jugador.getCasillaActual();
            vista.mostrar("Su posición actual es: \n" + casilla + "\n");
            boolean libre = !jugador.getEncarcelado();
            
            if (!libre){
                vista.mostrar("El jugador está en la cárcel :( \n");
                
                if (salirCarcel())
                    vista.mostrar("El jugador vuelve a ser libre \n");
                else
                    vista.mostrar("El jugador no ha conseguido salir de prisión \n");
            
                vista.mostrar("Pulsa \'intro\' para continuar");
                System.in.read();
            }
            
            if (libre){
                tienePropietario = juego.jugar();
                jugador = juego.getJugadorActual();
                casilla = jugador.getCasillaActual();
                vista.mostrar("Se tira el dado \n");
                vista.mostrar("El jugador " + jugador.getNombre() + " ha caído en la casilla \n" + casilla + "\n");
                vista.mostrar("Pulsa \'intro\' para continuar");
                System.in.read();
                
                eleccionSegunCasilla(casilla.getTipo(),tienePropietario);
                gestionInmobiliaria();
            
            }
            
            if (jugador.getSaldo() > 0){
                siguienteTurno();
                bancarrota = finJuego();
            }else
                bancarrota = finJuego();
        }
    }
    
    
    private boolean finJuego(){
        boolean fin = false;
        if (jugador.getSaldo()<=0){
            vista.mostrar(juego.obtenerRanking().toString());
            fin = true;
            vista.mostrar("El juego ha terminado \n");
        }
        return fin;
    }

    private void siguienteTurno(){
        jugador = juego.siguienteJugador();
    }
    
    private boolean salirCarcel(){
        int opcion = vista.menuSalirCarcel();
        boolean libre = false;
        MetodoSalirCarcel metodo = null;
        if (opcion == 0)
            metodo = MetodoSalirCarcel.TIRANDODADO;
        else if (opcion == 1)
            metodo = MetodoSalirCarcel.PAGANDOLIBERTAD;
        
        libre = juego.intentarSalirCarcel(metodo);
        return libre;
    }
    
    private void eleccionSegunCasilla(TipoCasilla tipo, boolean tienePropietario){
        
        if (jugador.getSaldo()>0 && !jugador.getEncarcelado()){
            if (casilla.getTipo() == TipoCasilla.CALLE){
                if(!tienePropietario){
                    if(vista.elegirQuieroComprar()){
                        if(juego.comprarTituloPropiedad()){
                            jugador = juego.getJugadorActual();
                            vista.mostrar("Se ha realizado la compra. Las propiedades del jugador son: \n");
                            vista.mostrar(jugador.getPropiedades().toString());
                        }else
                            vista.mostrar("No se ha podido comprar \n");
                    }
                }
  
            }else if (casilla.getTipo() == TipoCasilla.SORPRESA){
                tienePropietario = juego.aplicarSorpresa();
                jugador = juego.getJugadorActual();
                casilla = jugador.getCasillaActual();
                System.out.println("Se coge una carta sorpresa: " + juego.getCartaActual());
                System.out.println("La casilla actual es \n" + casilla + "\n");
                System.out.println("El estado del jugador tras la sorpresa es \n" + jugador + "\n");
                if (jugador.getSaldo()>0 && !jugador.getEncarcelado() && casilla.getTipo() == TipoCasilla.CALLE){
                    if(!tienePropietario){
                        if(vista.elegirQuieroComprar()){
                            if(juego.comprarTituloPropiedad()){
                                jugador = juego.getJugadorActual();
                                vista.mostrar("Se ha realizado la compra. Las propiedades del jugador son: \n");
                                vista.mostrar(jugador.getPropiedades().toString());
                            }else
                                vista.mostrar("No se ha podido comprar \n");
                        }
                    }
                }
            }
        }
    }
     
    private void gestionInmobiliaria(){
        if (!jugador.getEncarcelado() && jugador.getSaldo() > 0 && jugador.tengoPropiedades())
        {
            int opcion = -1;
            while(opcion != 0){
                opcion = vista.menuGestionInmobiliaria();
                if (opcion == 5){
                    ArrayList<Casilla> propiedades = juego.propiedadesHipotecadasJugador(true);
                    vista.mostrar(propiedades.toString());
                    casilla = elegirPropiedad(propiedades);
                }else if (opcion != 0){
                    ArrayList<Casilla> propiedades = juego.propiedadesHipotecadasJugador(false);
                    vista.mostrar(propiedades.toString());
                    casilla = elegirPropiedad(propiedades);
                }
                
            
                menuInmobiliario(opcion);
                jugador = juego.getJugadorActual();
                vista.mostrar("El estado del jugador tras la acción es: \n"); 
                vista.mostrar(jugador.toString());
            }
        }
    }
    
    private void menuInmobiliario(int opcion){
        switch (opcion) {
            case 1:
                if(juego.edificarCasa((Calle)casilla))
                    vista.mostrar("Se ha edificado la casa \n");
                else
                    vista.mostrar("No se ha podido edificar la casa \n");
                break;
            case 2:
                if(juego.edificarHotel((Calle)casilla))
                    vista.mostrar("Se ha edificado el hotel \n");
                else
                    vista.mostrar("No se ha edificado el hotel \n");
                break;
            case 3:
                if(juego.venderPropiedad((Calle)casilla))
                    vista.mostrar("Se ha vendido la propiedad \n");
                else
                    vista.mostrar("No se ha vendido la propiedad \n");
                break;
            case 4:
                if(juego.hipotecarPropiedad((Calle)casilla))
                    vista.mostrar("Se ha hipotecado la propiedad \n");
                else
                    vista.mostrar("No se ha podido hipotecar la propiedad \n");
                break;
            case 5:
                if(juego.cancelarHipoteca((Calle)casilla))
                    vista.mostrar("Se ha cancelado la hipoteca \n");
                else
                    vista.mostrar("No se ha podido cancelar la hipoteca \n");
                break;
            case 0:
                break; 
        }
    }
    
    public void inicializacionJuego() throws IOException{
        juego = Qytetet.getInstance();
        vista = new VistaTextualQytetet();
        ArrayList<String> nombres = new ArrayList();
        nombres = vista.obtenerNombreJugadores();
        juego.inicializarJuego(nombres);
        jugador = juego.getJugadorActual();
        casilla = jugador.getCasillaActual();
        System.out.println(juego);
        vista.mostrar("Pulsa \'intro\' para continuar");
        System.in.read();
    }
    
    private Casilla elegirPropiedad(ArrayList<Casilla> propiedades){ 
    //este método toma una lista de propiedades y genera una lista de strings, con el número y nombre de las propiedades
    //luego llama a la vista para que el usuario pueda elegir.
        vista.mostrar("\tCasilla\tTítulo");
        int seleccion;
        ArrayList<String> listaPropiedades= new ArrayList();
        for ( Casilla c: propiedades) {
            if(c != null){
                listaPropiedades.add( "\t"+c.getNumeroCasilla()+"\t"+((Calle) c).getTitulo().getNombre()); 
            }
        }
        seleccion=vista.menuElegirPropiedad(listaPropiedades);  
        return propiedades.get(seleccion);
    }
    
    public static void main(String[] args) throws IOException {
        
        ControladorQytetet cq = new ControladorQytetet();
        cq.inicializacionJuego();
        cq.desarrolloJuego();
    }
 
}

File: QytetetRuby/lib/controlador_qytetet.rb
#encoding: utf-8

module InterfazTextualQytetet
  require_relative "qytetet.rb"
  require_relative "vista_textual_qytetet.rb"
  require_relative "dado.rb"
  require_relative "jugador.rb"
  require_relative "sorpresa.rb"
  require_relative "tipo_sorpresa.rb"
  require_relative "tablero.rb"
  require_relative "casilla.rb"
  require_relative "tipo_casilla.rb"
  require_relative "titulo_propiedad.rb"
  require_relative "metodo_salir_carcel.rb"
  
  class ControladorQytetet
    
    def initialize
      @juego = nil
      @jugador = nil
      @casilla = nil
      @vista = VistaTextualQytetet.new
    end
  
    def desarrolloJuego
      bancarrota = false
      tienePropietario = false
        
      while (!bancarrota)
        @vista.mostrar("Es el turno de #{@jugador.getNombre} \n")
        @casilla = @jugador.getCasillaActual
        @vista.mostrar("Su posición actual es: #{@casilla} \n")
        libre = !@jugador.getEncarcelado
            
        if (!libre)
          @vista.mostrar("El jugador está en la cárcel :( \n")
                
          if (salirCarcel)
            @vista.mostrar("El jugador vuelve a ser libre \n")
          else
            @vista.mostrar("El jugador no ha conseguido salir de prisión \n")
          end
          @vista.mostrar("Pulsa \'intro\' para continuar")
          pausa
        end
            
        if (libre)
          tienePropietario = @juego.jugar
          @jugador = @juego.getJugadorActual
          @casilla = @jugador.getCasillaActual
          @vista.mostrar("Se tira el dado \n")
          @vista.mostrar("El jugador #{@jugador.getNombre} ha caído en la casilla #{@casilla} \n")
          @vista.mostrar("Pulsa \'intro\' para continuar")
          pausa
                
          eleccionSegunCasilla(@casilla.tipo,tienePropietario)
          gestionInmobiliaria
            
        end
            
        if (@jugador.getSaldo > 0)
          siguienteTurno
          bancarrota = finJuego
        else
          bancarrota = finJuego
        end
      end
    end
    
    
    def finJuego
      fin = false
      if (@jugador.getSaldo<=0)
        @vista.mostrar(@juego.obtenerRanking)
        fin = true
        @vista.mostrar("El juego ha terminado \n");
      end
      fin
    end

    def siguienteTurno
      @jugador = @juego.siguienteJugador
    end
    
    def salirCarcel
      opcion = @vista.menuSalirCarcel
      libre = false
      metodo = nil
      if (opcion == 0)
        metodo = ModeloQytetet::MetodoSalirCarcel::TIRANDODADO
      elsif (opcion == 1)
        metodo = ModeloQytetet::MetodoSalirCarcel::PAGANDOLIBERTAD
      end
        
      libre = @juego.intentarSalirCarcel(metodo)
      libre
    end
    
    def eleccionSegunCasilla(tipo, tienePropietario)
        
        if (@jugador.getSaldo>0 and !@jugador.getEncarcelado)
            if (@casilla.tipo == ModeloQytetet::TipoCasilla::CALLE)
              if(!tienePropietario)
                if(@vista.elegirQuieroComprar)
                  if(@juego.comprarTituloPropiedad)
                    @jugador = @juego.getJugadorActual
                    @vista.mostrar("Se ha realizado la compra \n")
                  else
                    @vista.mostrar("No se ha podido comprar \n")
                  end
                end
              end
            elsif (@casilla.tipo == ModeloQytetet::TipoCasilla::SORPRESA)
              tienePropietario = @juego.aplicarSorpresa
              @jugador = @juego.getJugadorActual
              @casilla = @jugador.getCasillaActual
              @vista.mostrar("Se coge una carta sorpresa: #{@juego.getCartaActual} \n")
              @vista.mostrar("La casilla actual es #{@casilla} \n")
              @vista.mostrar("El estado del jugador tras la sorpresa es #{@jugador} \n")
              if (@jugador.getSaldo>0 and !@jugador.getEncarcelado and @casilla.tipo == ModeloQytetet::TipoCasilla::CALLE)
                if(!tienePropietario)
                  if(@vista.elegirQuieroComprar)
                    if(@juego.comprarTituloPropiedad)
                      @jugador = @juego.getJugadorActual
                      @vista.mostrar("Se ha realizado la compra \n")
                      @vista.mostrar("#{@jugador.getPropiedades()}\n")
                    else
                      @vista.mostrar("No se ha podido comprar \n")
                    end
                  end
                end
              end
            end
        end
    end
     
    def gestionInmobiliaria
        if (!@jugador.getEncarcelado() and @jugador.getSaldo() > 0 and @jugador.tengoPropiedades())
          opcion = -1
          while(opcion != 0)
            opcion = @vista.menuGestionInmobiliaria
            if (opcion == 5)
              @casilla = elegirPropiedad(@juego.propiedadesHipotecadasJugador(true))
            elsif (opcion != 0)
              @casilla = elegirPropiedad(@juego.propiedadesHipotecadasJugador(false))
            end
            
            menuInmobiliario(opcion)
            @jugador = @juego.getJugadorActual
            @vista.mostrar("El estado del jugador tras la acción es #{@jugador}\n")
          end
        end
    end
    
    def menuInmobiliario(opcion)
      case opcion
        when 1
          @juego.edificarCasa(@casilla)
        when 2
          @juego.edificarHotel(@casilla)
        when 3
          @juego.venderPropiedad(@casilla)
        when 4
          @juego.hipotecarPropiedad(@casilla)
        when 5
          @juego.cancelarHipoteca(@casilla)
      else
        
      end
    end
    
    def inicializacionJuego
      @juego = ModeloQytetet::Qytetet.instance  
      nombres = Array.new
      nombres = @vista.obtener_nombre_jugadores
      @juego.inicializarJuego(nombres)
      @jugador = @juego.getJugadorActual
      @casilla = @jugador.getCasillaActual
      @vista.mostrar("#{@juego}")
    end
    
    def pausa
      puts "Pulsa \'intro\' para continuar"
      gets
    end
    
    #   # --------------------el siguiente método va en ControladorQytetet
    def elegir_propiedad(propiedades) # lista de propiedades a elegir
      @vista.mostrar("\tCasilla\tTitulo");
        
      listaPropiedades= Array.new
      for prop in propiedades  # crea una lista de strings con numeros y nombres de propiedades
        propString= prop.numeroCasilla.to_s+' '+prop.titulo.nombre; 
        listaPropiedades<<propString
      end
      seleccion=@vista.menu_elegir_propiedad(listaPropiedades)  # elige de esa lista del menu
      propiedades.at(seleccion)
    end
   
    
    def main
      inicializacionJuego
      desarrolloJuego
    end
 
  end
  c = ControladorQytetet.new
  c.main
end
